epm/Distribution/Client/IndexUtils.hs

```diff
@@ -450,7 +451,7 @@ packageIndexFromCache mkPkg hnd entrs mode = accum mempty [] entrs
       -- from the index tarball if it turns out that we need it.
       -- Most of the time we only need the package id.
       ~(pkg, pkgtxt) <- unsafeInterleaveIO $ do
-        mPatch <- patchedPackageCabalFile pkgid
+        mPatch <- patchedPackageCabalFile pkgid defaultPatchesDir
```

Comment:
Well I think the `Build Type` decides how to do the `configure` step so they need to know it up front even before extracting. My ignorance of this fact caused many frustrations.

There are other cases beyond testing where you would use the `--patches-directory` option. It could be incredibly useful if you just have a couple of tiny patches to add to existing libraries and you don't want to fork them.

I would certainly use this because when I'm building a production app, I can't wait for the maintainer to verify my patch and then upload a new version. The whole workflow is too slow for industrial use. This particular feature is needed for Hackage packages over other language ecosystems particularly because there are a lot of unmaintained packages on Hackage. The only packages I can think of that actually receive prompt responses are Michael Snoyman's. You could also fix cabal hell on the spot by removing upper version bounds and do other fixes necessary to make stuff work with your particular application.

So yeah, go ahead and finish the implementation, it's going to be a useful feature for end users as well. 

I'll file an issue for ignoring the custom build types for further notice.